content/2_Bronze/Ad_Hoc.mdx

```diff
@@ -35,6 +36,439 @@ approaching problems that appear to be ad hoc.
+
+This check takes $\mathcal{O}(N + M + K)$, which brings our total time complexity to $\mathcal{O}(N(N + M + K))$.
+
+<LanguageSection>
+<PySection>
+
+```py
+import sys
+
+sys.stdin = open("milkorder.in", "r")
+sys.stdout = open("milkorder.out", "w")
+
+n, m, k = map(int, input().split())
+
+hierarchy = [i - 1 for i in list(map(int, input().split()))]
+order = [-1] * n
+
+for i in range(k):
+	cow, pos = map(int, input().split())
+
+	order[pos - 1] = cow - 1
+
+	# already fixed, nothing we can do
+	if cow == 1:
+		print(pos)
+		exit(0)
+
+
+def check():
+	new_order = order[:]
+
+	cow_to_pos = [-1] * n
+	for i in range(n):
+		if order[i] != -1:
+			cow_to_pos[order[i]] = i
+
+	h_idx, i = 0, 0
+	while i < n and h_idx < m:
+		# we know the next cow has to be in front of it
+		if cow_to_pos[hierarchy[h_idx]] != -1:
+			if i > cow_to_pos[hierarchy[h_idx]]:
+				return False
+
+			i = cow_to_pos[hierarchy[h_idx]]
+			h_idx += 1
+		else:
+			while i < n and new_order[i] != -1:
+				i += 1
+
+			# run out of places
+			if i == n:
+				return False
+
+			new_order[i] = hierarchy[h_idx]
+			cow_to_pos[hierarchy[h_idx]] = i
+			h_idx += 1
+
+		i += 1
+
+	return True
+
+
+for i in range(n):
+	# if already fixed, skip
+	if order[i] == -1:
+		# try placing cow 1 @ position i
+		order[i] = 0
+
+		if check():
+			print(i + 1)
+			break
+
+		order[i] = -1
+```
+</PySection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int n, m, k;
+
+bool check(vector<int> order, vector<int> &hierarchy) {
+	vector<int> cow_to_pos(n, -1);
+
+	for (int i = 0; i < n; i++) {
+		if (order[i] != -1) { cow_to_pos[order[i]] = i; }
+	}
+
+	int h_idx = 0;
+	for (int i = 0; i < n && h_idx < m; i++) {
+		if (cow_to_pos[hierarchy[h_idx]] != -1) {
+			// we know the next cow has to be in front of it
+
+			if (i > cow_to_pos[hierarchy[h_idx]]) { return 0; }
+
+			i = cow_to_pos[hierarchy[h_idx]];
+			h_idx++;
+		} else {
+			while (i < n && order[i] != -1) { i++; }
+
+			// run out of places
+			if (i == n) { return 0; }
+
+			order[i] = hierarchy[h_idx];
+			cow_to_pos[hierarchy[h_idx]] = i;
+			h_idx++;
+		}
+	}
+
+	return 1;
+}
+
+int main() {
+	freopen("milkorder.in", "r", stdin);
+	freopen("milkorder.out", "w", stdout);
+	cin >> n >> m >> k;
+
+	vector<int> hierarchy(m);
+	for (int i = 0; i < m; i++) {
+		cin >> hierarchy[i];
+		hierarchy[i]--;
+	}
+
+	vector<int> order(n, -1);
+
+	for (int i = 0; i < k; i++) {
+		int cow, pos;
+		cin >> cow >> pos;
+
+		order[--pos] = --cow;
+
+		// already fixed, nothing we can do
+		if (cow == 0) {
+			cout << pos + 1 << endl;
+			return 0;
+		}
+	}
+
+	for (int i = 0; i < n; i++) {
+		// if already fixed, skip
+		if (order[i] == -1) {
+			// try placing cow 1 @ position i
+			order[i] = 0;
+			if (check(order, hierarchy)) {
+				cout << i + 1 << endl;
+				break;
+			}
+			order[i] = -1;
+		}
+	}
+}
+```
+</CPPSection>
+</LanguageSection>
+
+### Solution 2
```

Comment:
is this the internal sol
we can probably crap it if it's the case

Reply:
yeah i feel like we can cut this but
you do you

Reply:
hmmm it feels like kind of a waste cuz everything was already written up for it but sure

Reply:
i mean the code was hella scuffed too so